File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys


def lines(pc):
    count = 0
    for i in length:
        count += i // pc
    return count


k, n = map(int, sys.stdin.readline().split())
length = []
for _ in range(k):
    length.append(int(sys.stdin.readline()))

pl = 1
pr = max(length)

while pl <= pr:
    pc = (pl + pr) // 2
    if lines(pc) >= n:  # 만든 줄의 개수가 n개보다 많으면, 길이를 늘려 최대 길이로 만들 수 있기 때문에 pl = pc +1
        pl = pc + 1
    elif lines(pc) < n:  # 더 짧은 많은 개수를 만들기 위해 pr = pc -1
        pr = pc - 1

print(pr)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.js
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(filePath).toString().trim().split('\n');

// K개의 랜선 갖고 있음, 길이는 제각각
// N개의 같은 길이의 랜선으로 만들고 싶음
// N개보다 많이 만드는 것도 N개를 만드는 것에 포함된다.

const [k, target] = input[0].split(' ').map(Number);
const lines = [];
for (let i = 1; i <= k; i++) {
  lines.push(+input[i]);
}

let left = 0;
let right = Math.max(...lines);
let answer = 0;
while (left <= right) {
  let mid = Math.floor((left + right) / 2);

  let sum = lines.reduce((acc, cur) => {
    return acc + Math.floor(cur / mid);
  }, 0);

  if (sum < target) {
    right = mid - 1;
  } else {
    left = mid + 1;
    answer = mid;
  }
}

console.log(answer);
